diff --git a/Makefile.am b/Makefile.am
index ecb6e3bbe..d5b90d821 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -727,7 +727,7 @@ endif
 # library libags-vst
 if WITH_VST3
 libags_vst_la_CPPFLAGS = -std=gnu++2a -DDEVELOPMENT
-libags_vst_la_CXXFLAGS = $(CXXFLAGS) $(COMPILER_FLAGS) $(WARN_FLAGS) -O -I./ -I/usr/include/vst3 $(GLIB_CFLAGS) $(VST3_CXXFLAGS) $(W32API_CFLAGS)
+libags_vst_la_CXXFLAGS = $(CXXFLAGS) $(COMPILER_FLAGS) $(WARN_FLAGS) -O -I./ -I/app/include/vst3 $(GLIB_CFLAGS) $(VST3_CXXFLAGS) $(W32API_CFLAGS)
 libags_vst_la_LDFLAGS = $(LDFLAGS) $(VST3_LDFLAGS) -version-info 3:0:0 $(PIC_FLAGS) -pthread 
 libags_vst_la_LIBADD = -lm $(GLIB_LIBS) $(VST3_LIBS) $(W32API_LIBS)
 
diff --git a/ags/meson.build b/ags/meson.build
index 043d5b12d..d22a1444c 100644
--- a/ags/meson.build
+++ b/ags/meson.build
@@ -157,7 +157,7 @@ endif
 cxx_compiler_flags = [
   '-std=gnu++2a',
   '-DRELEASE',
-  '-I/usr/include/vst3'
+  '-I/app/include/vst3'
 ]
 
 libags_sources = files(
diff --git a/configure.ac b/configure.ac
index 71baf80bd..2899534fa 100644
--- a/configure.ac
+++ b/configure.ac
@@ -421,8 +421,8 @@ WITH_VST3=0
 
 AS_IF([test "x$enable_vst3" != xno],
       [AC_DEFINE([AGS_WITH_VST3], [1], [VST3 enabled])
-      VST3_CXXFLAGS="-I/usr/include/vst3"
-      VST3_LIBS="-L/usr/lib/vst3 -lbase -lpluginterfaces -lsdk -lsdk_common -lsdk_hosting -lvstgui -lvstgui_standalone -lvstgui_support -lvstgui_uidescription"      
+      VST3_CXXFLAGS="-I/app/include/vst3"
+      VST3_LIBS="-L/app/lib/vst3 -lbase -lpluginterfaces -lsdk -lsdk_common -lsdk_hosting -lvstgui -lvstgui_standalone -lvstgui_support -lvstgui_uidescription"      
       AC_SUBST(VST3_CXXFLAGS)
       AC_SUBST(VST3_LIBS)
       AC_MSG_NOTICE([VST3 enabled])],
